src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,232 @@
+          type,
+          start: new Position(info.line, info.indentation),
+          // Calculate position of last non-white character)
+          end: new Position(
+            endLine.line,
+            endLine.text.search(PythonDocument.reLastNonWhiteSpaceCharacter) - 1
+          ),
+        });
+      }
+    }
+
+    return structure;
+  }
+
+  /*
+   * Find the position of the specified:
+   *    type: function or class
+   *    direction: next or prev
+   *    edge: start or end
+   *
+   * With this information one can determine all of the required motions
+   */
+  find(type: Type, direction: Direction, edge: Edge, position: Position): Position | undefined {
+    // Choose the ordering method name based on direction
+    const isDirection = direction === 'next' ? 'isAfter' : 'isBefore';
+
+    // Filter function for all elements whose "edge" is in the correct "direction"
+    // relative to the cursor's position
+    const dir = (element: StructureElement) => element[edge][isDirection](position);
+
+    // Filter out elements from structure based on type and direction
+    const elements = this.structure.filter((elem) => elem.type === type).filter(dir);
+
+    if (edge === 'end') {
+      // When moving to an 'end' the elements should be started by the end position
+      elements.sort((a, b) => a.end.line - b.end.line);
+    }
+
+    // Return the first match if any exist
+    if (elements.length) {
+      // If direction === 'next' return the first element
+      // otherwise return the last element
+      const index = direction === 'next' ? 0 : elements.length - 1;
+      const element = elements[index];
+      const pos = element[edge];
+
+      // execAction MUST return a fully realized Position object created using new
+      return new Position(pos.line, pos.character);
+    }
+
+    return undefined;
+  }
+
+  // Use PythonDocument instance to move to specified class boundary
+  static moveClassBoundary(
+    document: TextDocument,
+    position: Position,
+    forward: boolean,
+    start: boolean
+  ): Position {
+    const direction = forward ? 'next' : 'prev';
+    const edge = start ? 'start' : 'end';
+
+    return new PythonDocument(document).find('class', direction, edge, position) || position;
+  }
+}
+
+type Type = 'function' | 'class';
+type Edge = 'start' | 'end';
+type Direction = 'next' | 'prev';
+
+// Uses the specified findFunction to execute the motion coupled to the shortcut (keys)
+abstract class BasePythonMovement extends BaseMovement {
+  abstract type: Type;
+  abstract direction: Direction;
+  abstract edge: Edge;
+
+  public async execAction(position: Position, vimState: VimState): Promise<Position> {
+    const document = vimState.document;
+    switch (document.languageId) {
+      case 'python':
+        return (
+          new PythonDocument(document).find(this.type, this.direction, this.edge, position) ||
+          position
+        );
+
+      default:
+        return position;
```

Comment:
If `document.languageId !== 'python'`, that's definitely a bug, right - since the caller is responsible for checking that? If that's the case, I'd rather throw an exception

Reply:
That is the part I am most uncertain about. Currently I am using [this code](https://github.com/abid-mujtaba/Vim/blob/b0f2b4d866a69dfa3134d92f7573ce95602322cd/src/actions/languages/python/motion.ts#L202-L208) to connect key-strokes to actions. The registration is not specifying a language so a user could, for example, press `]m` in a file of any language (not just Python).

What is the proper way to ensure that these key-stroke actions are only valid in a Python file?

Reply:
Oh, sorry, you're right - it's just `moveClassBoundary` that's called from outside. The right way to do this is to implement `doesActionApply`

Reply:
Done